File: solutions/clojure/allergies/2/src/allergies.clj
(ns allergies)

(def allergies-map
  { :eggs 1
    :peanuts 2
    :shellfish 4
    :strawberries 8
    :tomatoes 16
    :chocolate 32
    :pollen 64
    :cats 128 })

(def allergies-db
  [:eggs
   :peanuts
   :shellfish
   :strawberries
   :tomatoes
   :chocolate
   :pollen
   :cats])

(defn allergic-to? [score allergen]
  (-> allergen
      allergies-map
      (bit-and score)
      zero?
      not))

(defn allergies [score]
  (->> (range 0 8)
       (filter (partial bit-test score))
       (map allergies-db)))

File: solutions/zig/allergies/2/allergies.zig
const std = @import("std");
const EnumSet = std.EnumSet;

pub const Allergen = enum(u3) {
    eggs,
    peanuts,
    shellfish,
    strawberries,
    tomatoes,
    chocolate,
    pollen,
    cats,
};

pub fn isAllergicTo(score: u8, allergen: Allergen) bool {
    return score & @shlExact(@as(u8, 1), @intFromEnum(allergen)) != 0;
}

pub fn initAllergenSet(score: usize) EnumSet(Allergen) {
    var set = EnumSet(Allergen).init(.{});
    const allergens = [8]Allergen{ .eggs, .peanuts, .shellfish, .strawberries, .tomatoes, .chocolate, .pollen, .cats };
    for (allergens) |allergen| {
        if (score & (@as(usize, 1) << @intFromEnum(allergen)) != 0) {
            set.insert(allergen);
        }
    }
    return set;
}
